### ISSUE TITLE
Improve memory performance of overriding functions

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#19815]
I won't have a chance to tackle this, but I'd like to write it down nevertheless.

I was recently on a dinner with @nbp since Mozilla had a hackathon in Ljubljana.

We discussed security updates, but while doing so, Nicolas came up with the idea how to solve memory inefficiency of overrides.

### Current implementation

`.override` and friends keep a thunk (unevaluated expression) of the original function and call it with new arguments.

### Efficient implementation

`.override` and friends keep a filename from which the function can be imported, import it and then call it with new arguments.

---

It has to be thought through if this will yield some other issues, but it's worth a try.

**Comment by copumpkin** [comment]
I thought imported .nix files got "deduplicated" (basically kept around in memory) too.

**Comment by aneeshusa** [comment]
From my understanding, this only affects `override`, as `overrideDerivation` acts only on the resulting `derivation`, and [`overrideAttrs`](https://github.com/NixOS/nixpkgs/pull/18660) re-calls `mkDerivation` (which is used often/good to keep cached in memory), but neither by themselves keep a reference to the original function.

**Comment by cstrahan** [comment]
One thing that is highly desirable, IMO, is for overrides to "stack" (or be "composable"), so if we make any changes, I'd strongly suggest making sure we don't lose that feature.

**Comment by nbp** [comment]
@aneeshusa indeed, this is only for `override` and not `overridDerivation`.  I have other plans for `overrideDerivation` but these would be as heavy changes to nixpkgs.

@cstrahan This should not change this feature.  The idea is that instead of keeping the result of the `import path`, we would keep `path` instead.

**Comment by aneeshusa** [comment]
Off topic, but @nbp I'd like to chat with you (on IRC perhaps) about your planned changes to `overrideDerivation` - IMO most use cases are better served by `overrideAttrs` and I'd like to know what you're thinking.

**Comment by nbp** [comment]
@aneeshusa my plan for `overrideDerivation` is to completely remove it.  The only reason we need it is because we execute the `mkDerivation` function.  The idea would be to replace the `mkDerivation` call by the definition of one attribute inside the attribute set which is currently given as argument to `mkDerivation`.  With this scheme, `overrideDerivation` can be replaced by the update (`//`) operator.  Doing so will remove all the hunks that have to be kept-alive for potentially evaluating `overrideDerivation` on most derivation.

**Comment by copumpkin** [comment]
@nbp sort of like the technique at the end of this? http://r6.ca/blog/20140422T142911Z.html

**Comment by danbst** [comment]
The `//` doesn't support referencing to super attributes, so `overrideAttrs` would be still useful.

**Comment by nbp** [comment]
@danbst the super argument is the left-hand-side of the update operator, so if it does not have a name yet, you can probably name it before using the update operator on it.

**Comment by Profpatsch** [comment]
(triage) has this been improved since the last comments?
